src/goto-symex/symex_goto.cpp

```diff
@@ -513,6 +516,80 @@ void goto_symext::symex_goto(statet &state)
   }
 }
 
+void goto_symext::symex_unreachable_goto(statet &state)
+{
+  PRECONDITION(!state.reachable);
+  // This is like symex_goto, except the state is unreachable. We try to take
+  // some arbitrary choice that maintains the state guard in a reasonable state
+  // in order that it simplifies properly when states are merged (in
+  // merge_gotos). Note we can't try to evaluate the actual GOTO guard because
+  // our constant propagator might be out of date, since we haven't been
+  // executing assign instructions.
+
+  // If the guard is already false then there's no value in this state; just
+  // carry on and check the next instruction.
+  if(state.guard.is_false())
+  {
+    symex_transition(state);
+    return;
+  }
+
+  const goto_programt::instructiont &instruction = *state.source.pc;
+  PRECONDITION(instruction.is_goto());
+  goto_programt::const_targett goto_target = instruction.get_target();
+
+  auto queue_unreachable_state_at_target = [&]() {
+    framet::goto_state_listt &goto_state_list =
+      state.call_stack().top().goto_state_map[goto_target];
+    goto_statet new_state(state.guard_manager);
+    new_state.guard = state.guard;
+    new_state.reachable = false;
+    goto_state_list.emplace_back(state.source, std::move(new_state));
+  };
+
+  if(instruction.get_condition().is_true())
+  {
+    if(instruction.is_backwards_goto())
+    {
+      // Give up trying to salvage the guard:
+      state.guard.add(false_exprt());
+      symex_transition(state);
+    }
+    else
+    {
+      // Take the branch:
+      queue_unreachable_state_at_target();
+      state.guard.add(false_exprt());
+      symex_transition(state);
```

Comment:
the two last instructions are common to the if and else cases so this can be simplified